Implemented #343: Custom logic for filtering discovered plugins

antirotor opened the pull request:
### This implements #343 

Adding ability to filter plugins and modify plugins during discovery. This is using same `register_*` logic as already existing api. 

### Adding filter function:

```python
pyblish.api.register_discovery_filter(my_filter)
```
Filter function must return tuple `(plugin, filtered)` -> `(class, bool)`
**Plugin** is plugin class processed by filter function, **filtered** is bool indicating wheter plugin should be filtered out or not.

So simple filter function can look like:
```python
def my_filter_function(plugin):
    if plugin.__name__ == 'FilteredPlugin':
        return plugin, True
   return plugin, False
```
This can also modify plugin:
```python
def my_filter_function(plugin):
    if plugin.__name__ == 'ModifiedPlugin':
        plugin.optional = False
   return plugin, False
```

### This PR adds:
```
register_discovery_filter,
deregister_discovery_filter,
deregister_all_discovery_filters,
registered_discovery_filters
```

antirotor committed:
added ability to filter discovered plugins with callback function

antirotor committed:
fixed python 2 compatibility with tests

mottosso commented:
Whop, you are fast. Commented [here](https://github.com/pyblish/pyblish-base/issues/343#issuecomment-504491701).

Let's continue implementation-specifics here.

mottosso commented on pyblish/plugin.py:
```diff
@@ -31,12 +31,14 @@
 from .vendor import iscompatible, six
 
 log = logging.getLogger("pyblish.plugin")
+from pprint import pprint
```
Don't forget to remove unused imports.

antirotor committed:
fixed: filters now loop over plugins

antirotor commented:
So now basic filter example looks like this:

```python
def my_plugin_filter(plugins):
        for name, plugin in list(plugins.items()):
            if plugin.__name__ == "MyFilteredPlugin":
                del plugins[name]

        pass
```
Is like this because in python 3, you cannot delete from dictionary while iterating over it.

mottosso commented on pyblish/plugin.py:
```diff
@@ -1326,6 +1390,7 @@ def discover(type=None, regex=None, paths=None):
 
         plugins[plugin.__name__] = plugin
 
+    filter_plugins(plugins)
```
I think we should move this till the very last step; after conversion to a list, and after having been sorted. That would give the user final control over the items, including their order.

mottosso commented on pyblish/plugin.py:
```diff
@@ -1194,6 +1195,69 @@ def registered_targets():
+    filtered = False
+    for callback in _registered_plugin_filters:
+        try:
+            plugin, filtered = callback(plugins)
```
There's no need for the callback to return anything; it can change the items in-place, like they do in your test below.

mottosso commented:
@antirotor Thanks for this, would you mind if I made a few tweaks to your implementation?

mottosso committed:
Discovery filter tweaks
Lowered requirements for discovery filter (no return values) and enabled re-sorting of plug-ins from filter

mottosso commented:
Let me know what you think about these changes, and whether they still apply to your usecase.

antirotor commented:
Seems ok for me, looks cleaner. I have no problem with those changes.

mottosso committed:
Increment version

mottosso commented:
Great, then we're all set. :) Thanks @antirotor

mottosso commented:
Released as [1.8.0](https://github.com/pyblish/pyblish-base/releases/tag/1.8.0)
